File: assets/proof/sample.ts
/**
 * Proof corpus — TypeScript.
 * Exercises: doc comment, decorator, generic, enum, template literal,
 * regex, JSX-adjacent property access, async/await, optional chaining.
 */
import { readFile } from 'node:fs/promises';
import type { Result } from './result';

export enum Mode { Normal = 'normal', Insert = 'insert', Visual = 'visual' }

const CALLSIGN = /^[A-Z]{2}-\d{3}$/u;
const DEFAULT_THRUST = 0.82;

interface Airframe<T extends string = string> {
  readonly callsign: T;
  thrust: number;
  modes?: ReadonlyArray<Mode>;
}

@sealed
export class Fighter implements Airframe {
  static readonly ceiling = 65_000;
  #armed = false;

  constructor(public readonly callsign: string, public thrust = DEFAULT_THRUST) {
    if (!CALLSIGN.test(callsign)) throw new TypeError(`bad callsign: ${callsign}`);
  }

  get armed(): boolean { return this.#armed; }

  async load(path: string): Promise<Result<Airframe, Error>> {
    const raw = await readFile(path, 'utf8');
    const parsed = JSON.parse(raw) as Partial<Airframe>;
    return parsed?.callsign
      ? { ok: true, value: { ...parsed, thrust: parsed.thrust ?? DEFAULT_THRUST } as Airframe }
      : { ok: false, error: new Error('no callsign') };
  }
}

function sealed(target: Function): void { Object.seal(target.prototype); }

File: assets/proof/sample.go
// Proof corpus — Go.
// Exercises: struct tags, interfaces, goroutines, channels, error handling,
// builtin types, method receivers.
package airframe

import (
	"context"
	"errors"
	"fmt"
	"regexp"
)

const CeilingFt = 65_000

var callsignRe = regexp.MustCompile(`^[A-Z]{2}-\d{3}$`)

var ErrNoCallsign = errors.New("airframe: no callsign")

type Mode int

const (
	Normal Mode = iota
	Insert
	Visual
)

func (m Mode) String() string {
	switch m {
	case Normal:
		return "NORMAL"
	case Insert:
		return "INSERT"
	default:
		return "VISUAL"
	}
}

type Annunciator interface {
	Label() string
	Lit() bool
}

type Airframe struct {
	Callsign  string             `json:"callsign"`
	Thrust    float64            `json:"thrust,omitempty"`
	Telemetry map[string]float64 `json:"-"`
}

func New(ctx context.Context, callsign string) (*Airframe, error) {
	if !callsignRe.MatchString(callsign) {
		return nil, fmt.Errorf("%w: %q", ErrNoCallsign, callsign)
	}
	return &Airframe{Callsign: callsign, Thrust: 0.82, Telemetry: map[string]float64{}}, nil
}
